# 🔎 Análisis exploratorio: 🏃🏻‍➡️ Life Style Data 

## 1️⃣ **Entender el contexto del proyecto**

El análisis exploratorio que se va a realizar es sobre el dataset **Life Style Data**, el cuál puedes encontrar en el siguiente enlace:
🔗[https://www.kaggle.com/datasets/jockeroika/life-style-data](https://www.kaggle.com/datasets/jockeroika/life-style-data)

---

### 🎯 **Objetivo del análisis**

El propósito de este análisis es **explorar y comprender cómo los hábitos de vida, la alimentación y la actividad física influyen en distintos indicadores de salud y rendimiento físico**.  

A través de un análisis exploratorio detallado, se pretende:
- Identificar patrones generales en los datos relacionados con el estilo de vida de los participantes.  
- Evaluar la relación entre variables como la frecuencia de ejercicio, la dieta, el IMC y el consumo calórico.  
- Detectar posibles valores atípicos o inconsistencias en los datos que puedan afectar el modelado posterior.  
- Preparar el dataset para aplicar modelos predictivos que permitan **estimar indicadores de salud o rendimiento físico** (por ejemplo, las calorías quemadas o el porcentaje de grasa corporal).  

Además, en etapas posteriores se implementarán técnicas de **regresión lineal (simple y múltiple)**, **regularización (Ridge y Lasso)**, y **modelos de aprendizaje automático basados en árboles de decisión y ensambles (Random Forest y Gradient Boosting)** para comparar su rendimiento y capacidad predictiva.

En conjunto, este proyecto busca **extraer información útil y visualmente comprensible** sobre los factores que contribuyen a un estilo de vida saludable, apoyando la toma de decisiones y el diseño de rutinas personalizadas.

---

### 🧩 **Significado de las variables**

| **Columna** | **Descripción** |
|--------------|----------------|
| `Age` | Edad del participante (en años). |
| `Gender` | Género biológico (Masculino/Femenino). |
| `Weight (kg)` | Peso del individuo en kilogramos. |
| `Height (m)` | Altura del individuo en metros. |
| `Max_BPM` | Frecuencia cardíaca máxima registrada durante una sesión de ejercicio. |
| `Avg_BPM` | Frecuencia cardíaca promedio mantenida durante la sesión. |
| `Resting_BPM` | Frecuencia cardíaca en reposo antes de iniciar el entrenamiento. |
| `Session_Duration (hours)` | Duración de la sesión de ejercicio, en horas. |
| `Calories_Burned` | Total de calorías quemadas durante la sesión. |
| `Workout_Type` | Tipo de entrenamiento realizado (por ejemplo: fuerza, HIIT, cardio). |
| `Fat_Percentage` | Porcentaje de grasa corporal del individuo. |
| `Water_Intake (liters)` | Promedio diario de consumo de agua, en litros. |
| `Workout_Frequency (days/week)` | Número de días de entrenamiento por semana. |
| `Experience_Level` | Nivel de experiencia física (1 = Principiante, 2 = Intermedio, 3 = Avanzado). |
| `BMI` | Índice de Masa Corporal (IMC), medida de grasa corporal basada en peso y altura. |
| `Daily meals frequency` | Número de comidas consumidas diariamente. |
| `Physical exercise` | Indica el tipo o la frecuencia de la actividad física realizada. |
| `Carbs` | Consumo diario de carbohidratos (en gramos). |
| `Proteins` | Consumo diario de proteínas (en gramos). |
| `Fats` | Consumo diario de grasas (en gramos). |
| `Calories` | Total de calorías ingeridas a través de la alimentación diaria. |
| `meal_name` | Nombre de la comida (por ejemplo: desayuno, almuerzo, cena). |
| `meal_type` | Tipo de comida (por ejemplo: snack, plato principal, bebida). |
| `diet_type` | Tipo de dieta seguida (por ejemplo: cetogénica, vegana, equilibrada). |
| `sugar_g` | Contenido de azúcar en gramos por comida. |
| `sodium_mg` | Contenido de sodio en miligramos por comida. |
| `cholesterol_mg` | Contenido de colesterol en miligramos por comida. |
| `serving_size_g` | Tamaño de la porción del alimento en gramos. |
| `cooking_method` | Método de cocción utilizado (por ejemplo: hervido, frito, a la parrilla). |
| `prep_time_min` | Tiempo de preparación de la comida, en minutos. |
| `cook_time_min` | Tiempo de cocción de la comida, en minutos. |
| `rating` | Valoración de la comida o entrenamiento (escala de 1 a 5). |
| `Name of Exercise` | Nombre del ejercicio realizado. |
| `Sets` | Número de series completadas en el ejercicio. |
| `Reps` | Número de repeticiones por serie. |
| `Benefit` | Descripción del beneficio físico del ejercicio. |
| `Burns Calories (per 30 min)` | Calorías estimadas que se queman en 30 minutos de ese ejercicio. |
| `Target Muscle Group` | Grupo muscular principal trabajado en el ejercicio. |
| `Equipment Needed` | Equipamiento necesario para realizar el ejercicio. |
| `Difficulty Level` | Nivel de dificultad del ejercicio (Principiante, Intermedio, Avanzado). |
| `Body Part` | Parte del cuerpo principalmente involucrada (por ejemplo: brazos, piernas, pecho). |
| `Type of Muscle` | Tipo de músculo implicado (por ejemplo: superior, core, fuerza de agarre). |
| `Workout` | Nombre específico del entrenamiento o ejercicio. |

---

🗒️ *Fuente de los datos: Kaggle - Global Coffee Health Dataset (2024)*  

---

Lo primero de todo, será importar todas las librerías que vayamos a usar para realizar el análisis

In [63]:
import pandas as pd          # manejo de datos (tablas)
import numpy as np           # operaciones numéricas
import matplotlib.pyplot as plt  # gráficos
import seaborn as sns        # gráficos más bonitos
from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV
from sklearn.linear_model import LinearRegression, Ridge, Lasso
from sklearn.preprocessing import StandardScaler
from sklearn.tree import DecisionTreeRegressor, plot_tree
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

## 2️⃣ **Análisis descriptivo inicial**

Antes de comenzar con el análisis y la limpieza de datos, realizamos un **análisis descriptivo inicial** del dataset con el objetivo de conocer su estructura, tamaño y el tipo de información que contiene.

Este paso nos permite identificar el número de registros disponibles, las columnas existentes y obtener una primera impresión sobre la distribución general de los datos.

In [ ]:
# Cargar el dataset
df = pd.read_csv("data/life_style_data.csv")

Mostraremos la cantidad total de filas de nuestro dataset y todas las columnas que contiene para hacernos una idea de los datos que tenemos

In [ ]:
# Mostrar la cantidad total de filas y columnas del dataset
print(f"🌍 Total de filas del dataset: {len(df)}\n------------------------------------------------------------------------------")
print(df.columns)

🌍 Total de filas del dataset: 20000
------------------------------------------------------------------------------
Index(['Age', 'Gender', 'Weight (kg)', 'Height (m)', 'Max_BPM', 'Avg_BPM',
       'Resting_BPM', 'Session_Duration (hours)', 'Calories_Burned',
       'Workout_Type', 'Fat_Percentage', 'Water_Intake (liters)',
       'Workout_Frequency (days/week)', 'Experience_Level', 'BMI',
       'Daily meals frequency', 'Physical exercise', 'Carbs', 'Proteins',
       'Fats', 'Calories', 'meal_name', 'meal_type', 'diet_type', 'sugar_g',
       'sodium_mg', 'cholesterol_mg', 'serving_size_g', 'cooking_method',
       'prep_time_min', 'cook_time_min', 'rating', 'Name of Exercise', 'Sets',
       'Reps', 'Benefit', 'Burns Calories (per 30 min)', 'Target Muscle Group',
       'Equipment Needed', 'Difficulty Level', 'Body Part', 'Type of Muscle',
       'Workout', 'BMI_calc', 'cal_from_macros', 'pct_carbs', 'protein_per_kg',
       'pct_HRR', 'pct_maxHR', 'cal_balance', 'lean_mass_kg', 'exp

Mostramos también los datos que contienen las primeras 5 filas

In [ ]:
print("\n🔎 Primeras 5 filas del dataset:\n")
print(df.head())


📘 Información general del dataset:

     Age  Gender  Weight (kg)  Height (m)  Max_BPM  Avg_BPM  Resting_BPM  \
0  34.91    Male        65.27        1.62   188.58   157.65        69.05   
1  23.37  Female        56.41        1.55   179.43   131.75        73.18   
2  33.20  Female        58.98        1.67   175.04   123.95        54.96   
3  38.69  Female        93.78        1.70   191.21   155.10        50.07   
4  45.09    Male        52.42        1.88   193.58   152.88        70.84   

   Session_Duration (hours)  Calories_Burned Workout_Type  ...  \
0                      1.00          1080.90     Strength  ...   
1                      1.37          1809.91         HIIT  ...   
2                      0.91           802.26       Cardio  ...   
3                      1.10          1450.79         HIIT  ...   
4                      1.08          1166.40     Strength  ...   

   cal_from_macros  pct_carbs  protein_per_kg   pct_HRR  pct_maxHR  \
0          2139.59   0.500432        1.

Por último, mostramos las estadísticas básicas de cada columna (cantidad total de filas, mediana, mínimo, máximo, 25%, 50%, 70%)

In [84]:
print("\n📊 Estadísticas descriptivas:\n")
print(df.describe())


📊 Estadísticas descriptivas:

                Age   Weight (kg)    Height (m)       Max_BPM       Avg_BPM  \
count  20000.000000  20000.000000  20000.000000  20000.000000  20000.000000   
mean      38.851454     73.898832      1.723094    179.889702    143.704306   
std       12.114580     21.173010      0.127033     11.510805     14.267688   
min       18.000000     39.180000      1.490000    159.310000    119.070000   
25%       28.170000     58.160000      1.620000    170.057500    131.220000   
50%       39.865000     70.000000      1.710000    180.140000    142.990000   
75%       49.630000     86.100000      1.800000    189.425000    156.060000   
max       59.670000    130.770000      2.010000    199.640000    169.840000   

        Resting_BPM  Session_Duration (hours)  Calories_Burned  \
count  20000.000000              20000.000000     20000.000000   
mean      62.195813                  1.259446      1280.109601   
std        7.289351                  0.341336       502.228

## 3️⃣ **Detección y tratamiento datos ausentes**

Antes de proceder con el análisis exploratorio y la creación de nuevas variables, es fundamental **verificar la existencia de valores nulos o ausentes** en el dataset.  

La presencia de datos faltantes puede afectar la precisión de los resultados, alterar las estadísticas descriptivas y comprometer el rendimiento de los modelos predictivos.  

Por ello, en esta sección se realiza una revisión general para comprobar si existen columnas o registros con valores nulos que deban ser tratados o imputados.

In [68]:
total_nulos = df.isnull().sum()
print(total_nulos)

Age                               0
Gender                            0
Weight (kg)                       0
Height (m)                        0
Max_BPM                           0
Avg_BPM                           0
Resting_BPM                       0
Session_Duration (hours)          0
Calories_Burned                   0
Workout_Type                      0
Fat_Percentage                    0
Water_Intake (liters)             0
Workout_Frequency (days/week)     0
Experience_Level                  0
BMI                               0
Daily meals frequency             0
Physical exercise                 0
Carbs                             0
Proteins                          0
Fats                              0
Calories                          0
meal_name                         0
meal_type                         0
diet_type                         0
sugar_g                           0
sodium_mg                         0
cholesterol_mg                    0
serving_size_g              

Tras ejecutar el análisis, se observa que ninguna de las columnas contiene valores nulos.

Esto indica que el dataset está completo y no requiere imputación ni eliminación de registros, por lo que se puede continuar con el análisis exploratorio sin necesidad de realizar tratamientos adicionales sobre los datos faltantes.

## 4️⃣ **Revisar los tipos de las variables**

Una vez verificada la integridad de los datos, el siguiente paso consiste en **revisar los tipos de las variables** para asegurarnos de que cada columna tenga el formato adecuado según el tipo de información que representa.

Este proceso es fundamental para evitar errores en los cálculos posteriores y garantizar que las operaciones estadísticas o de modelado se apliquen correctamente.  

In [69]:
# Mostramos un resumen de los tipos de datos de cada columna
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 54 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   Age                             20000 non-null  float64
 1   Gender                          20000 non-null  object 
 2   Weight (kg)                     20000 non-null  float64
 3   Height (m)                      20000 non-null  float64
 4   Max_BPM                         20000 non-null  float64
 5   Avg_BPM                         20000 non-null  float64
 6   Resting_BPM                     20000 non-null  float64
 7   Session_Duration (hours)        20000 non-null  float64
 8   Calories_Burned                 20000 non-null  float64
 9   Workout_Type                    20000 non-null  object 
 10  Fat_Percentage                  20000 non-null  float64
 11  Water_Intake (liters)           20000 non-null  float64
 12  Workout_Frequency (days/week)   

En este caso, hemos detectado algunas columnas que con formato decimal cuando deberían ser un número entero, por lo que convertiremos esas columnas para el correcto tratamiento de los datos.

Las columnas que vamos a convertir a número entero son: Workout_Frequency, Experience_level y Daily meals frequency.

El motivo para convertir estas columnas entero, son que, por ejemplo en el caso de `Workout_Frequency`, lo que representa es el número de días que la persona entrena por semana, no tiene sentido entrenar 3.6 días a la semana.

Por otro lado, en el caso de `Experience_Level` lo que representa es el nivel de experiencia física (1 = Principiante, 2 = Intermedio, 3 = Avanzado), son etiquéticas numéricas, por lo que no tiene sentido que sean float.

Por último, la variable `Daily meal frequency` representa el número de comidas, al igual que ocurre con `Workout_Frequency`, es uin conteo de comidas, 4 comidas, 5 comidas, no tiene sentido que nadie coma 4.3 veces al día.

In [70]:
df['Workout_Frequency'] = df['Workout_Frequency (days/week)'].astype(int)
df['Experience_Level'] = df['Experience_Level'].astype(int)
df['Daily meals frequency'] = df['Daily meals frequency'].astype(int)

df.head()

Age  Gender  Weight (kg)  Height (m)  Max_BPM  Avg_BPM  Resting_BPM  \
0  34.91    Male        65.27        1.62   188.58   157.65        69.05   
1  23.37  Female        56.41        1.55   179.43   131.75        73.18   
2  33.20  Female        58.98        1.67   175.04   123.95        54.96   
3  38.69  Female        93.78        1.70   191.21   155.10        50.07   
4  45.09    Male        52.42        1.88   193.58   152.88        70.84   

   Session_Duration (hours)  Calories_Burned Workout_Type  ...  pct_carbs  \
0                      1.00          1080.90     Strength  ...   0.500432   
1                      1.37          1809.91         HIIT  ...   0.500850   
2                      0.91           802.26       Cardio  ...   0.500610   
3                      1.10          1450.79         HIIT  ...   0.499533   
4                      1.08          1166.40     Strength  ...   0.500581   

   protein_per_kg   pct_HRR  pct_maxHR  cal_balance  lean_mass_kg  \
0        1.624789  0.741237   0.835985       725.10     47.777394   
1        1.514093  0.551247   0.734270      -232.91     40.809803   
2        1.663445  0.574534   0.708124       805.74     44.635580   
3        0.862017  0.744155   0.811150      1206.21     63.007432   
4        2.538153  0.668405   0.789751       303.60     43.347504   

   expected_burn  Burns Calories (per 30 min)_bc  Burns_Calories_Bin  \
0       685.1600                    7.260425e+19              Medium   
1       978.6184                    1.020506e+20                High   
2       654.5266                    1.079607e+20                High   
3       773.6300                    8.987921e+19                High   
4       711.4176                    5.264685e+19                 Low   

   Workout_Frequency  
0                  3  
1                  4  
2                  2  
3                  3  
4                  4  

[5 rows x 55 columns]

## 5️⃣ **Identificación y tratamiento de datos atípicos**

En esta sección se lleva a cabo la identificación y análisis de posibles valores atípicos presentes en el dataset. Los outliers pueden deberse a errores de registro, inconsistencias en la carga de datos o simplemente reflejar casos extremos pero reales dentro de la población.

El objetivo es verificar la coherencia y plausibilidad de los datos, asegurando que las observaciones sean consistentes con la realidad del contexto físico y nutricional que se analiza.

Para este proceso se utilizó principalmente el método `describe()`, que permite obtener estadísticas descriptivas básicas como los valores `mínimos`, `máximos`, `medias`, `medianas` y `percentiles`. Gracias a ello, se pudieron detectar valores que inicialmente podrían parecer extremos.

No obstante, en lugar de eliminarlos directamente, se realizó una revisión contextual comparando variables relacionadas —por ejemplo, la relación entre el peso, la altura y las calorías quemadas durante una sesión— para determinar si los valores eran coherentes desde un punto de vista fisiológico.

Este enfoque permite distinguir entre errores reales y casos válidos, evitando eliminar información relevante y manteniendo la integridad del dataset para el análisis posterior.

In [71]:
# Usamos la función describe para evaluar de forma rápida si hay algún valor atípico en nuestro dataset
df.describe()

Age   Weight (kg)    Height (m)       Max_BPM       Avg_BPM  \
count  20000.000000  20000.000000  20000.000000  20000.000000  20000.000000   
mean      38.851454     73.898832      1.723094    179.889702    143.704306   
std       12.114580     21.173010      0.127033     11.510805     14.267688   
min       18.000000     39.180000      1.490000    159.310000    119.070000   
25%       28.170000     58.160000      1.620000    170.057500    131.220000   
50%       39.865000     70.000000      1.710000    180.140000    142.990000   
75%       49.630000     86.100000      1.800000    189.425000    156.060000   
max       59.670000    130.770000      2.010000    199.640000    169.840000   

        Resting_BPM  Session_Duration (hours)  Calories_Burned  \
count  20000.000000              20000.000000     20000.000000   
mean      62.195813                  1.259446      1280.109601   
std        7.289351                  0.341336       502.228982   
min       49.490000                  0.490000       323.110000   
25%       55.960000                  1.050000       910.800000   
50%       62.200000                  1.270000      1231.450000   
75%       68.090000                  1.460000      1553.112500   
max       74.500000                  2.020000      2890.820000   

       Fat_Percentage  Water_Intake (liters)  ...  cal_from_macros  \
count    20000.000000           20000.000000  ...     20000.000000   
mean        26.101313               2.627486  ...      1998.297076   
std          4.996160               0.604724  ...       440.848408   
min         11.333134               1.460000  ...      1105.570000   
25%         22.387807               2.170000  ...      1661.022500   
50%         25.822504               2.610000  ...      1943.130000   
75%         29.676026               3.120000  ...      2271.950000   
max         35.000000               3.730000  ...      3699.540000   

          pct_carbs  protein_per_kg       pct_HRR     pct_maxHR   cal_balance  \
count  20000.000000    20000.000000  20000.000000  20000.000000  20000.000000   
mean       0.499983        1.460142      0.699005      0.802305    744.308699   
std        0.001455        0.518946      0.144880      0.096613    720.946619   
min        0.492434        0.516706      0.371344      0.599789  -1266.220000   
25%        0.499054        1.076294      0.583656      0.727676    261.432500   
50%        0.499981        1.382260      0.686284      0.794834    691.190000   
75%        0.500910        1.750495      0.798196      0.869211   1176.290000   
max        0.507889        3.916881      1.073939      1.047032   3075.580000   

       lean_mass_kg  expected_burn  Burns Calories (per 30 min)_bc  \
count  20000.000000   20000.000000                    2.000000e+04   
mean      53.786384     866.352318                    8.631802e+19   
std       12.498740     250.317069                    3.197579e+19   
min       30.946261     219.852800                    2.491905e+16   
25%       44.587037     714.098250                    6.441978e+19   
50%       51.204908     868.721400                    8.371578e+19   
75%       61.939016    1012.532700                    1.100442e+20   
max       90.117371    1477.108800                    1.756614e+20   

       Workout_Frequency  
count       20000.000000  
mean            2.922700  
std             1.033382  
min             1.000000  
25%             2.000000  
50%             3.000000  
75%             4.000000  
max             5.000000  

[8 rows x 40 columns]

Si analizamos los datos que nos reporta el método `describe()`, observamos algunos valores que en un primer momento podrían parecer atípicos, como el máximo de calorías quemadas por sesión (`Calories_Burned` = 2890.82 kcal).

Sin embargo, al revisar conjuntamente las columnas de altura y peso, comprobamos que los valores máximos registrados (`Height` = 2.01 m, `Weight` = 130.77 kg) corresponden a individuos con una complexión grande.

Esto tiene sentido desde el punto de vista fisiológico, ya que una persona más alta y con mayor masa corporal consume más energía durante el ejercicio, incluso realizando la misma actividad que alguien más pequeño.

Además, el análisis de la variable `Experience_Level` muestra que la mayoría de participantes se encuentran entre los niveles 1 y 2 (principiante e intermedio), lo que sugiere que el gasto calórico elevado puede deberse también a menor eficiencia en la ejecución del ejercicio, ya que las personas con menos experiencia tienden a gastar más energía para mantener la misma intensidad.

Por tanto, estos valores altos de calorías no se consideran errores ni outliers anómalos, sino variaciones naturales explicadas por las características físicas y el nivel de experiencia de los individuos.

---

Durante la exploración inicial del dataset, se identificaron algunas columnas cuyos valores no coincidían con la descripción oficial proporcionada por la fuente del conjunto de datos, o bien no aportaban información relevante para el análisis.
Este proceso de revisión tiene como objetivo garantizar la calidad y consistencia semántica de las variables, eliminando aquellas que puedan inducir ruido o confusión en los resultados posteriores.

En concreto:

- La columna `meal_name` se eliminó porque todos sus valores corresponden al término "Other", a pesar de que la descripción oficial indica que debería contener nombres de comidas como Breakfast, Lunch o Dinner. Al no aportar variabilidad ni información útil, esta variable se descartó.

- La columna `Physical exercise` fue removida tras comprobar que sus valores numéricos (por ejemplo, -0.04, 0.97, 3.0) no se ajustaban a la descripción original (“indica el tipo o la frecuencia de actividad física”), lo que sugiere un error en la codificación o carga de datos.

Estas eliminaciones permiten mantener la coherencia entre el significado teórico de las variables y los datos reales presentes en el dataset, mejorando la fiabilidad del análisis posterior.

In [72]:
print((f"Total columnas antes de la limpieza: {len(df.columns)}"))

# Creamos un nuevo DataFrame sin las columnas que no son útiles
df = df.drop(['meal_name','Physical exercise'], axis=1)

# Mostramos de nuevo las 4 primeras filas para verificar que está todo correcto
print((f"Total columnas después de la limpieza: {len(df.columns)}"))

Total columnas antes de la limpieza: 55
Total columnas después de la limpieza: 53


## 6️⃣ **Creación de variables derivadas o auxiliares**

Tras la fase de limpieza y verificación de datos, hemos decidido **crear una serie de variables auxiliares** para enriquecer el análisis y obtener una visión más completa sobre los factores relacionados con el **rendimiento físico, la nutrición y la condición corporal** de los participantes.  

El objetivo principal de este apartado es **aprovechar la información ya existente** en el dataset para generar nuevos indicadores que nos ayuden a interpretar mejor los resultados, detectar patrones y facilitar el análisis de correlaciones en fases posteriores.  

Estas variables combinan datos de distintas columnas originales y nos permiten comparar de manera más justa el desempeño entre personas con características físicas diferentes. A continuación, se describen las nuevas variables creadas:

- **`HRR (Heart Rate Reserve)`**: calcula la **reserva de frecuencia cardíaca**, es decir, la diferencia entre la frecuencia máxima y la frecuencia en reposo (`Max_BPM - Resting_BPM`). Este valor refleja la capacidad cardiovascular de cada individuo: cuanto mayor sea el HRR, mejor condición física suele tener la persona.

- **`Relative intensity`**: mide el **nivel de esfuerzo relativo** durante la sesión utilizando la **fórmula de Karvonen**, un método ampliamente utilizado para calcular la intensidad del ejercicio en función de la frecuencia cardíaca individual. Se obtiene mediante la expresión  **`(Avg_BPM - Resting_BPM) / HRR`**,  donde `HRR` representa la reserva de frecuencia cardíaca. Este valor refleja el **porcentaje de la capacidad cardiovascular utilizada** durante la sesión: cuanto más alto es el resultado, mayor es la intensidad del esfuerzo realizado en relación con el máximo potencial del individuo.

- **`Calories/hour`**: representa la **intensidad del entrenamiento** expresada en calorías quemadas por hora (`Calories_Burned / Session_Duration (hours)`). Esta métrica permite comparar el esfuerzo entre sesiones de distinta duración de manera más equitativa.

- **`Macronutrient ratios`**: muestra la **proporción de carbohidratos, proteínas y grasas** respecto al total de calorías ingeridas. Para calcularlo se asume que los carbohidratos y proteínas aportan 4 kcal/g y las grasas 9 kcal/g. Estos porcentajes ayudan a entender la calidad nutricional de la dieta de cada individuo.

- **`Weekly training load`**: estima la **carga total de entrenamiento semanal**, combinando la frecuencia y la duración media de las sesiones (`Workout_Frequency * Session_Duration`). Cuanto mayor sea este valor, mayor es el volumen de entrenamiento acumulado en la semana.

- **`Approximate energy balance`**: mide el **balance energético diario aproximado** restando las calorías quemadas de las calorías ingeridas (`Calories - Calories_Burned`). Un valor positivo indica superávit calórico (ingesta mayor al gasto), mientras que un valor negativo indica déficit calórico.

La creación de estas variables nos permite **profundizar en el análisis exploratorio y predictivo**, ofreciendo una visión más contextualizada del rendimiento físico y los hábitos de los participantes. Gracias a estos nuevos indicadores, podremos identificar relaciones más precisas entre el entrenamiento, la dieta y los resultados fisiológicos.

In [ ]:
# Reserva de frecuencia cardíaca (FCR)
df['HRR'] = df['Max_BPM'] - df['Resting_BPM']

# Intensidad relativa
# Usamos la fómrula de Karvonen que es un método eficaz para calcular nuestras zonas de entrenamiento por frecuencia cardíaca
df['Relative_Intensity'] = (df['Avg_BPM'] - df['Resting_BPM']) / df['HRR']

# Calorías por hora (intensidad de entrenamiento)
df['Calories_per_hour'] = df['Calories_Burned'] / df['Session_Duration (hours)'].replace(0, np.nan)

# Ratio de macronutrientes (%)
# Calculamos las calorías estimadas por macros
macro_kcal = (df['Carbs'] * 4) + (df['Proteins'] * 4) + (df['Fats'] * 9)

df['Carb_ratio'] = (df['Carbs'] * 4 / macro_kcal).round(3)
df['Protein_ratio'] = (df['Proteins'] * 4 / macro_kcal).round(3)
df['Fat_ratio'] = (df['Fats'] * 9 / macro_kcal).round(3)

# Carga de entrenamiento semanal
df['Weekly_training_load'] = df['Workout_Frequency (days/week)'] * df['Session_Duration (hours)']

# Balance energético aproximado
df['Energy_balance'] = df['Calories'] - df['Calories_Burned']

# Verificación rápida de las nuevas columnas
df[['HRR', 'Relative_Intensity', 'Calories_per_hour',
    'Carb_ratio', 'Protein_ratio', 'Fat_ratio',
    'Weekly_training_load', 'Energy_balance']].head()

HRR  Relative_Intensity  Calories_per_hour  Carb_ratio  Protein_ratio  \
0  119.53            0.741237        1080.900000       0.500          0.198   
1  106.25            0.551247        1321.102190       0.501          0.200   
2  120.08            0.574534         881.604396       0.501          0.200   
3  141.14            0.744155        1318.900000       0.500          0.199   
4  122.74            0.668405        1080.000000       0.501          0.200   

   Fat_ratio  Weekly_training_load  Energy_balance  
0      0.301                3.9900          725.10  
1      0.300                5.4800         -232.91  
2      0.300                2.7209          805.74  
3      0.302                4.3890         1206.21  
4      0.299                4.3200          303.60

## 7️⃣ **Análisis de correlaciones**

En esta sección se analizan las relaciones entre distintas variables del dataset con el objetivo de responder a preguntas específicas sobre la salud, el rendimiento físico y los hábitos de los participantes en este dataset.

### 🩺 Bloque A: Salud cardiovascular

**Objetivo:** Entender cómo los hábitos de vida afectan el rendimiento y la eficiencia del corazón.

**Contexto:** El ritmo cardíaco en reposo es un indicador clave de la salud cardiovascular y del estado físico general. Factores como la hidratación, la frecuencia de entrenamiento, la intensidad relativa del ejercicio y la calidad de la dieta pueden influir directamente en la eficiencia del sistema cardiovascular.  

Analizar estas relaciones permite comprender si los participantes con mejores hábitos presentan una mejor capacidad cardiovascular reflejada en un menor `Resting_BPM` y un mayor `HRR`.

**Pregunta 1:** ¿Las personas con hábitos más saludables (mayor consumo de agua, frecuencia de entrenamiento y dieta equilibrada) presentan menor `Resting_BPM` y mayor `HRR`?  
**Pregunta 2:** ¿A mayor frecuencia de entrenamiento, menores son las pulsaciones en reposo?

**📊 Análisis sugerido:**
- Calcular correlaciones entre `Water_Intake`, `Workout_Frequency`, `Relative_Intensity` y las variables cardiovasculares (`Resting_BPM`, `HRR`).  
- Visualizar relaciones con gráficos de dispersión o regresión lineal.  
- Posible uso de `sns.heatmap()` o `sns.regplot()` para observar las correlaciones.


**Conclusión:** (Interpretación clara del resultado)

### 💪 Bloque B: Rendimiento de sesión


**Objetivo:** Evaluar qué factores influyen en el gasto energético y la intensidad del entrenamiento.

**Contexto:** El rendimiento durante una sesión de ejercicio depende de múltiples variables, como el tipo de entrenamiento, la experiencia del individuo y su capacidad cardiovascular. Analizar estas relaciones permite identificar qué factores determinan una mayor quema de calorías, si los entrenamientos más intensos (como HIIT) se traducen realmente en un gasto calórico superior, y si los niveles de experiencia afectan la eficiencia del esfuerzo realizado.

**Pregunta 1:** ¿Qué tipo de ejercicio tiende a quemar más calorías? (`Workout_Type` vs `Calories_Burned`)  
**Pregunta 2:** ¿Las personas con menos experiencia (`Experience_Level`) queman más calorías por sesión que las avanzadas?  

**📊 Análisis sugerido:**
- Gráficos de barras (`sns.barplot`) comparando el tipo de entrenamiento con las calorías quemadas y la intensidad.  
- Cálculo de medias mediante `groupby` y `mean()` para `Calories_Burned` y `Calories/hour`.  
- Comparaciones de medias entre distintos niveles de experiencia (`Experience_Level`).
- Posible visualización con `sns.boxplot()` para ver la variabilidad dentro de cada tipo de entrenamiento.

**Conclusión:** (Interpretación clara del resultado)

### 🍽️ Bloque C: Hábitos alimenticios y composición corporal


**Objetivo:**  Evaluar la relación entre la dieta y el estado físico general de los participantes.

**Contexto:**  La alimentación es un pilar fundamental en la salud y el rendimiento físico. Analizar cómo la ingesta calórica total y la distribución de macronutrientes se relacionan con indicadores como el `BMI` o el `Energy_balance` permite determinar si las dietas seguidas por los participantes son equilibradas o presentan desequilibrios.

Además, entender qué tipos de dietas son más frecuentes y qué proporciones de macronutrientes predominan ayuda a interpretar los patrones nutricionales del conjunto de datos.

**Pregunta 1:** ¿Las dietas con mayor ingesta calórica se asocian con un `BMI` más alto o más bajo?  
**Pregunta 2:** ¿Qué proporciones de macronutrientes son más comunes entre los participantes?  

**📊 Análisis sugerido:**  
- Calcular la correlación entre `Calories` y `BMI`.  
- Visualizar las distribuciones de `Carb_ratio`, `Protein_ratio` y `Fat_ratio` mediante histogramas o boxplots.  
- Comparar los porcentajes de macronutrientes medios por tipo de dieta (`diet_type`).  


**Conclusión:** (Interpretación clara del resultado)

### ⚖️ Bloque D: Perfil general de los participantes

**Objetivo:** Comprender el perfil dominante en el dataset y analizar las diferencias entre grupos.

**Contexto:** Explorar las variables demográficas y de estilo de vida permite identificar el perfil general de los individuos del dataset: si son más bien sedentarios o activos, si entrenan con regularidad y si su balance energético es positivo o negativo.

Además, observar las diferencias entre hombres y mujeres en cuanto a frecuencia de entrenamiento, tipo de ejercicios y gasto calórico puede revelar tendencias interesantes sobre los hábitos de entrenamiento y alimentación.

**Pregunta 1:** ¿Predominan sujetos con un estilo de vida activo o sedentario?  
**Pregunta 2:** ¿Entrenan con más frecuencia los hombres o las mujeres? (`Gender` vs `Workout_Frequency`)  
**Pregunta 3:** ¿Qué tipo de ejercicio prefieren los hombres y las mujeres? (`Gender` vs `Workout_Type`)  
**Pregunta 4:** ¿La mayoría de los sujetos están en superávit o déficit calórico? (`Energy_balance`)

**📊 Análisis sugerido:**  
- Gráficos de barras y conteo (`sns.countplot`, `sns.barplot`) para comparar grupos por género o hábitos.  
- Cálculo de estadísticas descriptivas de `Workout_Frequency` por `Gender`.  
- Histograma de `Energy_balance` para visualizar la distribución entre déficit y superávit calórico.  
- Comparación de medias y medianas para determinar diferencias significativas entre grupos.

**Conclusión:** (Interpretación clara del resultado)

## 8️⃣ **Modelado predictivo**

## 9️⃣ **Regularización**

## 🔟 **Árboles de regresión y ajuste de hiperparámetros**

## 1️⃣1️⃣ **Random Forest vs Gradient Boosting**

## 📢 **Conclusiones finales**